CSV Parser (#8254)
DataDog/documentation: content/en/logs/processing/parsing.md

| key:"/valueStr"              | `%{data::keyvalue(":", "/")}`                         | {"key": "/valueStr"}                  |
| /key:/valueStr               | `%{data::keyvalue(":", "/")}`                         | {"/key": "/valueStr"}                 |
| key:={valueStr}              | `%{data::keyvalue(":=", "", "{}")}`                   | {"key": "valueStr"}                   |

> | key1=value1\|key2=value2     | <code>%{data::keyvalue(&quot;=&quot;, &quot;&quot;, &quot;&quot;, &quot;&#124;&quot;)}</code> | {"key1": "value1", "key2": "value2"}  |

Reviewer: ```suggestion
| key1=value1\|key2=value2     | `%{data::keyvalue(&quot;=&quot;, &quot;&quot;, &quot;&quot;, &quot;&#124;&quot;)}` | {"key1": "value1", "key2": "value2"}  |
```

Reviewer: Backticks for parity - for find & replace situations in the future

> | key1="value1"\|key2="value2" | <code>%{data::keyvalue(&quot;=&quot;, &quot;&quot;, &quot;&quot;, &quot;&#124;&quot;)}</code> | {"key1": "value1", "key2": "value2"}  |

Reviewer: ```suggestion
| key1="value1"\|key2="value2" | `%{data::keyvalue(&quot;=&quot;, &quot;&quot;, &quot;&quot;, &quot;&#124;&quot;)}` | {"key1": "value1", "key2": "value2"}  |
```

Author: Actually I'll revert this and keep the <code> as it's the only working work around I found which allows me to put a `|` in a code section within a table 😢 
hack found on https://stackoverflow.com/questions/17319940/how-to-escape-a-pipe-char-in-a-code-statement-in-a-markdown-table

* If the XML contains tags that have both an attribute and a sting value between the two tags, a `value` attribute is generated. For example: `<title lang="en">Harry Potter</title>` is converted to `{"title": {"lang": "en", "value": "Harry Potter" } }`
* Repeated tags are automatically converted to arrays. For example: `<bookstore><book>Harry Potter</book><book>Everyday Italian</book></bookstore>` is converted to `{ "bookstore": { "book": [ "Harry Potter", "Everyday Italian" ] } }`

### Parsing CSV

> Use the **csv** filter to more-easily map strings to attributes when separated by a given character (`,` by default).

Reviewer: ```suggestion
Use the **CSV** filter to more-easily map strings to attributes when separated by a given character (`,` by default).
```

> The csv filter is defined as `csv(headers[, separator[, quotingcharacter]])` where:

Reviewer: ```suggestion
The CSV filter is defined as `csv(headers[, separator[, quotingcharacter]])` where:
```

> * `headers`: defines the keys name separated by `,`. Keys names must start with alphabetical character and can contain any alphanumirical character in addition to `-`, `_`, `@` and `$`.

Reviewer 2: For the keys:
The first character of a column must be a letter, an underscore (_). Subsequent characters of a column must be a letter, a digit or an underscore (_).

Reviewer 3: Indeed, `@` and `$` are not currently supported. Also, it's actually restricted to **ascii** alphabetic characters.

> * `headers`: defines the keys name separated by `,`. Keys names must start with alphabetical character and can contain any alphanumirical character in addition to `-`, `_`, `@` and `$`.

Reviewer: ```suggestion
* `headers`: Defines the keys name separated by `,`. Keys names must start with alphabetical character and can contain any alphanumerical character in addition to `-`, `_`, `@` and `$`.
```

> * `headers`: defines the keys name separated by `,`. Keys names must start with alphabetical character and can contain any alphanumirical character in addition to `-`, `_`, `@` and `$`.

Reviewer: ```suggestion
* `headers`: Defines the keys name separated by `,`. Keys names must start with alphabetical character and can contain any alphanumerical character in addition to `-`, `_`, `@` and `$`.
```

Reviewer 3: Shouldn't it be `alphanumeric` instead of `alphanumerical`?

> * `separator`: defines separators used to separate the different values. Only one character is accepted. Default: `,` . Note use `tab` to represent the tabulation character.

Reviewer: ```suggestion
* `separator`: defines separators used to separate the different values. Only one character is accepted. Default: `,` . **Note**: Use `tab` to represent the tabulation character.
```

> * `quotingcharacter`: defines the quoting character. Only one character is accepted. Default: `"` 

Reviewer: ```suggestion
* `quotingcharacter`: Defines the quoting character. Only one character is accepted. Default: `"` 
```

**Note**:

> * Values containing a separator character must but quoted.

Reviewer 3: ```suggestion
* Values containing a separator character must be quoted.
```

> * Quoted Values containing a quoting character must be escaped with a quoting characters. ex: `""` represent `"`.

Reviewer: ```suggestion
* Quoted Values containing a quoting character must be escaped with a quoting characters. For example, `""` represent `"`.
```

Author: Done

> * Quoted Values containing a quoting character must be escaped with a quoting characters. ex: `""` represent `"`.

Reviewer 3: ```suggestion
* Quoted Values containing a quoting character must be escaped with a quoting characters. ex: `""` within a quoted value represents `"`.
```

> * If the log doesn't contain the same number of value as the number of keys in the header, the csv parser will match the first ones. 

Reviewer: ```suggestion
* If the log doesn't contain the same number of value as the number of keys in the header, the CSV parser will match the first ones. 
```

**Log**:

> ```text

Reviewer: We've switched to new code blocks so users have an easy copy button / better view! :)

{{< code-block lang="json" filename="block.json" >}}
your code example here
{{< /code-block >}}

Author: Done

> John,Doe,120 jefferson st.,Riverside, NJ, 08075

Reviewer: ```suggestion
John,Doe,120 Jefferson St.,Riverside, NJ, 08075
```

> John,Doe,120 jefferson st.,Riverside, NJ, 08075

Reviewer 3: Although this example correctly shows the current behavior, I would avoid the whitespace after the last comma, as this is a specific case where our CSV parser has potentially questionable behavior (we strip whitespace outside of quotes, which is different than a lot of other CSV parsers). Having a more "standard" example CSV here allows us to avoid documenting this behavior for now.

```suggestion
John,Doe,120 jefferson st.,Riverside,NJ,08075
```

Author: Added in the Notes and in the first example.

```

**Rule**:

```text
myParsingRule %{data:user:csv("first_name,name,adress,city,state,postal_code")}
```

**Result:**

  ```json
{
  "user": {
    "first_name": "John",
    "name": "Doe",

>     "adress": "120 jefferson st.",

Reviewer: ```suggestion
    "address": "120 Jefferson St.",
```

    "city": "Riverside",
    "state": "NJ",
    "postal_code": "08075"
  }
}
  ```

Other examples:

> | **Raw string**               | **Parsing rule**                                                         | **Result**                                      |

Reviewer 3: It's probably worth mentioning (or showing) that numbers and boolean values are parsed into actual numbers and booleans (instead of their string representation)

Author: Added in the Notes and in the first example

|:-----------------------------|:-------------------------------------------------------------------------|:------------------------------------------------|
| `John,Doe`                   | `%{data::csv("firstname,name")}`                                         | {"firstname": "John", "name":"Doe"}             |
| `"John ""Da Man""",Doe`      | `%{data::csv("firstname,name")}`                                         | {"firstname": "John \"Da Man\"", "name":"Doe"}  |
| `'John ''Da Man''',Doe`      | `%{data::csv("firstname,name",",","'")}`                                 | {"firstname": "John 'Da Man'", "name":"Doe"}    |

> | <code>John&#124;Doe</code>   | <code>%{data::csv(&quot;firstname,name&quot;,&quot;&#124;&quot;)}</code> | {"firstname": "John", "name":"Doe"}             |

Reviewer: ```suggestion
| <code>John&#124;Doe</code>   | `%{data::csv(&quot;firstname,name&quot;,&quot;&#124;&quot;)}` | {"firstname": "John", "name":"Doe"}             |
```

Reviewer: Backticks for parity - in case we ever have to do a find and replace situation in the future
